### PR TITLE
[build] Remove the legacy loading mode.

### PR DESCRIPTION
**Pull request by ejgallego**
This is WIP, but reached the point for a first submission.

Generated dune files contain correct paths, however as these paths are
referring to the `install` directory, dune fails.

We do a ugly fixup, I guess we could do better.

Note that Dune's ML API doesn't have this problem, thus:

```
$ cp theories/dune.disabled theories/dune
$ dune build theories
```
works fine!

### CODE REVIEW

**Comment by SkySkimmer** [comment]
What's the needed independent fix? Also I think the dependency got merged.

**Comment by ejgallego** [comment]
@SkySkimmer the required independent fix is that coqdep doesn't properly resolve the findlib paths in windows, due to drive letters not playing nice with the relativization path. See Windows CI job.

File: user-contrib/Ltac2/Init.v [hunk 1]
```diff
@@ -8,8 +8,9 @@
 (*         *     (see LICENSE file for the text of the license)         *)
 (************************************************************************)
 
-Declare ML Module "ltac2_plugin:coq-core.plugins.ltac2".
-Declare ML Module "ltac2_ltac1_plugin:coq-core.plugins.ltac2_ltac1".
+(* EJGA: Seems that Coq's findlib loader is not loading this correctly? *)
+Declare ML Module "coq-core.plugins.ltac2".
```
Comment (ejgallego) [inline user-contrib/Ltac2/Init.v:12]: It is quite strange that without this, the build fails. Findlib should load the plugin automatically?

Generated rule in the failing case is:
```sexp
(rule
 (targets Init.timing Init.glob Init.vos Init.vo)
 (deps ../../theories/Init/Prelude.vo
       %{project_root}/_build/install/default/lib/coq-core/META
       %{project_root}/_build/install/default/lib/coq-core/plugins/ltac2/ltac2_plugin.cmxs
       %{project_root}/_build/install/default/lib/coq-core/META
       %{project_root}/_build/install/default/lib/coq-core/plugins/ltac2_ltac1/ltac2_ltac1_plugin.cmxs)
 (action (chdir ../../. (run coqc -boot -R theories Coq -Q user-contrib/Ltac2 Ltac2 -w +default -q -w -deprecated-native-compiler-option -native-compiler off -time-file user-contrib/Ltac2/Init.timing %{dep:Init.v}))))
```
which looks fine to me, META file looks fine too.

**Comment by ejgallego** [comment]
Other than the windows build problem, I think this is looking good.

It would be still good to fix the problem with plugins that extend the grammar needing to be loaded manually instead of via findlib.

Debug output is:
```
[DEBUG] Fl_dynload: not loading: unix
[DEBUG] Fl_dynload: not loading: threads.posix
[DEBUG] Fl_dynload: not loading: threads
[DEBUG] Fl_dynload: not loading: coq-core.config
[DEBUG] Fl_dynload: not loading: coq-core.boot
[DEBUG] Fl_dynload: not loading: str
[DEBUG] Fl_dynload: not loading: coq-core.clib
[DEBUG] Fl_dynload: not loading: coq-core.perf
[DEBUG] Fl_dynload: not loading: coq-core.lib
[DEBUG] Fl_dynload: not loading: coq-core.gramlib
[DEBUG] Fl_dynload: not loading: coq-core.vm
[DEBUG] Fl_dynload: not loading: dynlink
[DEBUG] Fl_dynload: not loading: coq-core.kernel
[DEBUG] Fl_dynload: not loading: coq-core.library
[DEBUG] Fl_dynload: not loading: coq-core.engine
[DEBUG] Fl_dynload: not loading: coq-core.pretyping
[DEBUG] Fl_dynload: not loading: zarith
[DEBUG] Fl_dynload: not loading: coq-core.interp
[DEBUG] Fl_dynload: not loading: coq-core.parsing
[DEBUG] Fl_dynload: not loading: coq-core.proofs
[DEBUG] Fl_dynload: not loading: coq-core.printing
[DEBUG] Fl_dynload: not loading: coq-core.tactics
[DEBUG] Fl_dynload: not loading: findlib.internal
[DEBUG] Fl_dynload: not loading: findlib
[DEBUG] Fl_dynload: not loading: findlib.dynload
[DEBUG] Fl_dynload: not loading: coq-core.vernac
[DEBUG] Fl_dynload: not loading: coq-core.plugins.ltac
[DEBUG] Fl_dynload: about to load: coq-core.plugins.ltac2
[DEBUG] Fl_dynload: files="ltac2_plugin.cmxs"
[DEBUG] Fl_dynload: loading "ltac2_plugin.cmxs"
[DEBUG] Fl_dynload: about to load: coq-core.plugins.ltac2_ltac1
[DEBUG] Fl_dynload: files="ltac2_ltac1_plugin.cmxs"
[DEBUG] Fl_dynload: loading "ltac2_ltac1_plugin.cmxs"
File "./user-contrib/Ltac2/Init.v", line 13, characters 0-49:
Error:
Anomaly
"Uncaught exception Failure("Grammar.extend: No top level in entry \"ltac2_atom\"")."
Please report at http://coq.inria.fr/bugs/.
```

File: user-contrib/Ltac2/Init.v (hunk 1 above)
Comment (ejgallego) [reply user-contrib/Ltac2/Init.v:12]: *update* this is due to `Declare ML Module` having to do some extra magic for plugins that extend the grammar.

We should document that better, I'm open to suggestions.

**Comment by ejgallego** [comment]
Ideally we could reimplement `Fl_dynload.load_packages` ourselves, it is not so hard, and add the corresponding grammar hooks for every plugin loaded.

File: user-contrib/Ltac2/Init.v (hunk 1 above)
Comment (SkySkimmer) [reply user-contrib/Ltac2/Init.v:12]: (Grammar update is the most common one but in general Declare ML Module runs the plugin_init_functions)

Comment (ejgallego) [reply user-contrib/Ltac2/Init.v:12]: I'm trying to research if there is a way to have findlib run that kind of stuff for us, I guess we could actually have the plugins register the pending initialization stuff to a table, then we process it after Dynload.

Anyways we can document this restriction for now, let's try to see if we can this patch in like that, it is only a minor annoyance.

Comment (ejgallego) [reply user-contrib/Ltac2/Init.v:12]: > I'm trying to research if there is a way to have findlib run that kind of stuff for us, I guess we could actually have the plugins register the pending initialization stuff to a table, then we process it after Dynload.

However that may be tricky w.r.t. summaries, so OMMV.

**Comment by ejgallego** [comment]
Let's restore parsing of the old Declare ML Form, as not to force the overlays so much.

This will however delay fixing the problem about some findlib names not being accepted by us.

**Comment by ejgallego** [comment]
@coqbot run full CI

**Comment by ejgallego** [comment]
@coqbot run full CI

**Comment by Alizter** [comment]
@ejgallego This doesn't appear to work with HoTT's Dune setup. Could you check what we are missing or if something needs to change here/

**Comment by ejgallego** [comment]
@Alizter the status of this PR is a bit complex as indeed here we are in a build situation that is sadly very non-deterministic due to path relativization. To sum up:

- in some build contexts, for example Coq's CI, path relativiation works fine and the build succeeds
- in other build contexts, path relativization goes to the root
- it is not clear what absolute paths make dune choke (c.f. https://github.com/ocaml/dune/issues/5833 )

Best would be of course to fix the dune issue so we don't need to do this path hackery.

I think for our case, the simplest fix for now could be to reject to do path relativization if the paths don't have any common prefix, but I am not sure there are more wrong cases.

Also, the path relativization code is broken on windows.

**Comment by SkySkimmer** [comment]
How can we finish this? Is the windows bug reported or explained somewhere?

**Comment by ppedrot** [comment]
What can we do? I think this is a high-priority PR that nobody knows how to make progress on.

**Comment by ejgallego** [comment]
Moreover my previous answer for Ali sums the situation up pretty well.

**Comment by ppedrot** [comment]
>  I know how to make progress so I'm unsure what you mean by nobody?

Your summary seems to indicate there is a mysterious dune bug involved. This is a high-priority PR, somebody ought to do something about it *quickly*.

**Comment by SkySkimmer** [comment]
What is path relativization? Where is it implemented? What is wrong with it?

**Comment by SkySkimmer** [comment]
>By far, the biggest priority for the build system is to merge Ali's port of the test suite to dune.

>That alone would save days of work to the dev team per month.

That sounds exaggerated to me, and I'm happy not having to deal with the bad errors it looked like it would produce.

**Comment by SkySkimmer** [comment]
>By the way, in terms of build systems roadmap, this PR is of priority wishlist.

It's important because whenever we deal with plugins like in https://github.com/coq/coq/pull/19313 the legacy stuff gets in the way

**Comment by ejgallego** [comment]
> What is path relativization? Where is it implemented? What is wrong with it?

It is a procedure in coqdep that turns the absolute path returned by findlib into a relative path.

**Comment by ejgallego** [comment]
> > By the way, in terms of build systems roadmap, this PR is of priority wishlist.
> 
> It's important because whenever we deal with plugins like in #19313 the legacy stuff gets in the way

How does the legacy stuff get in the way?

**Comment by SkySkimmer** [comment]
There's a bunch of random hacks to remember, as you can see the PR still doesn't work

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
The pushed commit (I also rebased) seems to work

File: Makefile [hunk 2]
```diff
@@ -127,6 +127,7 @@ DUNESTRAPOPT=--root .
 # We regenerate always as to correctly track deps, can do better
 # We do a single call to dune as to avoid races and locking
 _build/default/theories_dune _build/default/ltac2_dune .dune-stamp: FORCE
+	$(if $(COQ_DUNE_EXTRA_OPT),echo '()' > dunestrap.inc)
```
Comment (SkySkimmer) [inline Makefile:130]: not sure this is the best way to work, basically in split mode we shouldn't be depending on the local coq-core META

**Comment by herbelin** [comment]
Just by curiosity, if ever someone has time for an answer, what are the advantages and disavantages of the new and old loading mode?

**Comment by Alizter** [comment]
@herbelin IIRC the new loading mode is better because it uses the findlib name of the library being loaded instead of the local cmxs. This means findlib is able to find the library better. In Dune speak, we switched from using the private library name (which isn't very meaningful) to the public library name.

**Comment by ejgallego** [comment]
> There's a bunch of random hacks to remember, as you can see the PR still doesn't work

I see the PR is merged, but indeed I still don't understand why this PR would help with that too much. Basically now you have to:
- do a `.mlpack` file
- correctly register the dependencies
- register the plugin in some random table

1 and 3 would be not necessary with this PR, however 2 which is more critical is still not solved due to gramlib stuff. 1 and 3 are easy boilerplate, so indeed, once that is documented properly, they take 1 min to do.

File: Makefile (hunk 2 above)
Comment (ejgallego) [reply Makefile]: Why not? Note that the generation of the `META` file is (or was) O(1) and doesn't build anything else.

File: tools/coqdep/lib/fl.ml [hunk 3]
```diff
@@ -107,7 +107,12 @@ let findlib_resolve ~meta_files ~file ~package ~plugin_name =
     in
     List.fold_right add path file
   in
-  let meta_file = normalize_path (to_relative_path meta_file) in
+  let meta_file =
+    match Sys.getenv_opt "DUNE_SOURCEROOT" with
```
Comment (ejgallego) [inline tools/coqdep/lib/fl.ml:111]: Can we maybe add a comment to this?

**Comment by ejgallego** [comment]
> The pushed commit (I also rebased) seems to work

Good, indeed only real blocker was the path normalization stuff.

We need to try if things still work with regular composed builds etc... even if not critical that would be nice to see.

I am pretty confident things will be fine, so we should be able to go ahead.

File: Makefile (hunk 2 above)
Comment (SkySkimmer) [reply Makefile]: When we use `-p coq-stdlib` dune says "no rule for ..." for stuff in coq-core

**Comment by rlepigre** [comment]
> Just by curiosity, if ever someone has time for an answer, what are the advantages and disavantages of the new and old loading mode?

To add to @Alizter's answer @herbelin, it is basically impossible to work with a plugin that has OCaml dependencies with the old loading mode. There are workarounds, like making a custom `.cmxs` that embeds all dependencies, but that is not really practical. Another option is to create "dummy plugins" for all the dependencies, but then you need to figure out what is the right order to load them in manually (and there are extra issues with dependencies whose findlib name is of the form `package` and not `package.component`).

File: tools/coqdep/lib/common.ml [hunk 4]
```diff
@@ -126,30 +111,30 @@ module VCache = Set.Make(VData)
...
     on ".v" files (for [Load]) or ".cmx", ".cmo", etc... (for [Declare]). *)
 
-let legacy_mapping = Core_plugins_findlib_compat.legacy_to_findlib
+let warn_legacy_loading =
+  let name = "legacy-loading-removed-coqdep" in
+  CWarnings.create ~name (fun name ->
+      Pp.(str "Legacy loading plugin method has been removed from Coq, \
+               and the `:` syntax is deprecated, and its first \
+               argument ignored; please remove \"" ++
+          str name ++ str ":\" from your Declare ML"))
```
Comment (herbelin) [inline tools/coqdep/lib/common.ml:120]: Will it be clear what the terminology "legacy loading plugin method" means? Would there be a way to describe it without using a word whose meaning is intrinsically unstable over time.

**Comment by herbelin** [comment]
@Alizter, @rlepigre: I see, thanks for the explanations.

File: Makefile [hunk 5]
```diff
@@ -127,6 +127,9 @@ DUNESTRAPOPT=--root .
 # We regenerate always as to correctly track deps, can do better
 # We do a single call to dune as to avoid races and locking
 _build/default/theories_dune _build/default/ltac2_dune .dune-stamp: FORCE
+ifneq ($(filter -split,$(COQ_DUNE_EXTRA_OPT)),) # avoid depending on local coq-core
```
Comment (SkySkimmer) [inline Makefile:130]: Maybe instead of changing the dependencies of _build/default/theories_dune depending on if we are split or not, we should have a separate target _build/default/theories_dune_split? The makefile can then pick which one it builds and copies out of _build.

File: tools/coqdep/lib/fl.ml (hunk 3 above)
Comment (SkySkimmer) [reply tools/coqdep/lib/fl.ml:113]: added a little comment

File: tools/coqdep/lib/common.ml (hunk 4 above)
Comment (SkySkimmer) [reply tools/coqdep/lib/common.ml:120]: It seems clear from context that it means the one using `:` no?

File: tools/coqdep/lib/fl.ml (hunk 3 above)
Comment (ejgallego) [reply tools/coqdep/lib/fl.ml:113]: Thanks, will look into this PR tomorrow, it's been a while so I need to refocus on this; but IIRC we should be able to get this merged as-is.

Comment (SkySkimmer) [reply tools/coqdep/lib/fl.ml:113]: Not sure if the normalization here is useful anymore. When called through gen_rules it will get fed into fixup_install_locations anyway, so fixup_install_locations could handle it.
Not sure about when it's called in a regular dune-coq project though.

File: tools/coqdep/lib/common.ml (hunk 4 above)
Comment (herbelin) [reply tools/coqdep/lib/common.ml:120]: We can reconstruct from the context that there is indeed an history of loading methods and that one of them used `:'`.

I'm playing the candid here, but in which sense (from the user point of view), is the new loading method essentially a new method rather than just a modification of the method?

Depending on the answer, a possible suggestion is:
```suggestion
      Pp.(strbrk "Declare ML Module now ignores the \"" ++ \
          str name ++ str ":\" part of its argument; \
          this part is deprecated, please remove it."))
```

File: Makefile (hunk 5 above)
Comment (SkySkimmer) [reply Makefile]: changed

**Comment by ppedrot** [comment]
Is this ready? I'm definitely not an expert of the build system, but I believe this is on the critical path of a reasonable handling of plugins. Since @SkySkimmer is a coauthor of the PR, in case of need I can assign myself.

**Review by Alizter — DISMISSED** [review]
The overlay proposed in https://github.com/HoTT/Coq-HoTT/pull/1951 doesn't build and I'm not sure why. I don't think we should progress with this until we understand what has happened there.

TLDR: Dune has a "bug" where paths to symlinks are not normalised "correctly" https://github.com/ocaml/dune/issues/5833. This is worked around in coqdep using https://github.com/coq/coq/pull/18165. This works fine for Coq's CI since coq is in the project root, however when Coq appears outside of the project root the coqdep normalization doesn't return the correct thing causing the build in https://github.com/HoTT/Coq-HoTT/pull/1951 to fail.

**Comment by ejgallego** [comment]
There are two things here that IMHO require attention:
- The way plugin dependency works should be documented better, for instance, it is not the case that if plugin `B` depends on pluging `A`, we can rely on findlib to load `A` correctly, hence users still have the obligation to declare all the explicit Coq plugin dependencies.
- The Super Hack that this PR adds by rewriting paths should be removed before merging.

**Comment by Alizter** [comment]
TLDR: Dune has a "bug" where paths to symlinks are not normalised "correctly" https://github.com/ocaml/dune/issues/5833. This is worked around in coqdep using https://github.com/coq/coq/pull/18165. This works fine for Coq's CI since coq is in the project root, however when Coq appears outside of the project root the coqdep normalization doesn't return the correct thing causing the build in https://github.com/HoTT/Coq-HoTT/pull/1951 to fail.

File: tools/dune_rule_gen/coq_rules.ml [hunk 6]
```diff
@@ -217,6 +171,11 @@ module Context = struct
 
 end
 
+(* Super hack, to be removed *)
+let fixup_install_locations file =
```
Comment (ejgallego) [inline tools/dune_rule_gen/coq_rules.ml:175]: This seems very suspicious and should not be needed.

**Comment by SkySkimmer** [comment]
I see https://github.com/coq-community/coq-plugin-template/ still uses legacy syntax `Declare ML Module "my_plugin:coq-my-plugin.plugin".`
Should it get an update?

**Comment by ejgallego** [comment]
This will also break jsCoq; as there has been some interest in maintaining it, I signal it here.

**Comment by ejgallego** [comment]
> This will also break jsCoq; as there has been some interest in maintaining it, I signal it here.

jsCoq and coq-lsp have been now updated to the new loading mode.

File: tools/dune_rule_gen/coq_rules.ml (hunk 6 above)
Comment (SkySkimmer) [reply tools/dune_rule_gen/coq_rules.ml]: I think I figured it out, cf latest commit

Comment (SkySkimmer) [reply tools/dune_rule_gen/coq_rules.ml]: (I tested dev mode and opam install locally)

Comment (SkySkimmer) [reply tools/dune_rule_gen/coq_rules.ml]: I guess I did something weird when testing dev mode because now it doesn't work

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: tools/dune_rule_gen/coq_rules.ml (hunk 6 above)
Comment (SkySkimmer) [reply tools/dune_rule_gen/coq_rules.ml]: another attempt 74119b4078

File: tools/dune_rule_gen/path.ml [hunk 7]
```diff
@@ -10,7 +10,25 @@ type t = Rel of string | Abs of string
...
+  if Filename.is_relative path then
+    (* when called by dune, the cmxs and META files are in
+       ../install/... relative to cwd (= _build/default)
+       but the generated dune file will be moved to ../theories
+       so adjusting the path to be relative to the .v won't work
+
+       (it would be relative to the .v in
+       project_root/_build/default/theories but dune would read it as
+       relative to the .v in project_root/theories, the number of ..
+       to insert to get to project_root doesn't match) *)
```
Comment (SkySkimmer) [inline tools/dune_rule_gen/path.ml:23]: I think this discrepancy between the location when generated and the location when used is the main reason why we need to do some adjustment

(not sure if this file is the right place to do the adjustment)

**Comment by Alizter** [comment]
I don't know the best way to test it right now, but I am working on it.

Even if it does fix the problem, I would suggest leaving this coqdep fix in for two versions so that the overlay https://github.com/HoTT/Coq-HoTT/pull/1951 becomes backwards comaptible when the legacy loading method is removed.

**Comment by SkySkimmer** [comment]
> I would suggest leaving this coqdep fix in for two versions

I don't understand what that means.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@Alizter I pushed to the overlay, AFAICT it is backwards compatible now

**Comment by ppedrot** [comment]
IIUC this is now ready to merge. Is everybody fine with the current state? @Alizter @ejgallego

**Comment by SkySkimmer** [comment]
The doc needs a bit of work imo

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@ejgallego can you open PRs for the overlays?

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 8]
```diff
@@ -659,12 +659,14 @@ file is a particular case of a module called a *library file*.
...
-   Loads an OCaml plugin and its dependencies dynamically.  The :n:`@string`
-   argument must be a valid `findlib <http://projects.camlcity.org/projects/findlib.html>`_
-   plugin name, for example ``coq-core.plugins.ltac``. As of Coq 8.16,
-   the command also supports a legacy
-   syntax compatible with the plugin loading system used in Coq
-   8.0-8.15, see below.
+   Loads OCaml plugins and their dependencies dynamically.  The :n:`@string`
+   arguments must be valid `findlib <http://projects.camlcity.org/projects/findlib.html>`_
+   plugin names, for example ``coq-core.plugins.ltac``.
+   They must also contain a `.` to disambiguate against the removed non-findlib loading method.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:665]: What happens if they don't contain a `.`?  Do you get an error message or the user has to remember?

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 9]
```diff
@@ -659,12 +659,14 @@ file is a particular case of a module called a *library file*.
...
+
+   Effects (such as adding new commands) from the explicitly requested
+   plugin are activated, but effects from implicitly loaded
+   dependencies are not activated.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:669]: This is vague.

"effects" means adding or changing commands and tactics (and nothing else)?  If so, maybe say that and leave out "effect".  "effect" is vague, it basically just means a change.  Dictionary definition:

> Effect is usually a noun meaning "a change that results when something is done or happens,"

Also, what does "activated" mean here?  How do you later activate implicitly loaded dependencies--you need a separate `Declare ML Module`?  If so, say more explicitly.


How do you subsequently activate

**Review by jfehrle — COMMENTED** [review]
wording suggestions

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 8 above)
Comment (SkySkimmer) [reply doc/sphinx/proof-engine/vernacular-commands.rst:665]: error

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 9 above)
Comment (SkySkimmer) [reply doc/sphinx/proof-engine/vernacular-commands.rst:669]: >"effect" is vague

plugins can do vague things so it's accurate

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 8 above)
Comment (rlepigre) [reply doc/sphinx/proof-engine/vernacular-commands.rst:665]: Can't this actually be relaxed with this MR, given that there does not seem to be anything to disambiguate anymore?

Comment (ppedrot) [reply doc/sphinx/proof-engine/vernacular-commands.rst:665]: I propose that we defer this change to another PR. It's not currently used, meanwhile the legacy loading mechanism has passed its shelf life by a large amount. The sooner we get rid of it the better, and the attempted branch date for 9.0 is coming close.

**Comment by ppedrot** [comment]
Ah, but the overlay PRs have not been opened yet... Good thing that I missed the merge.

**Comment by rlepigre** [comment]
> Ah, but the overlay PRs have not been opened yet... Good thing that I missed the merge.

Seems like they are opened but not linked?
- https://github.com/mattam82/Coq-Equations/pull/592
- https://github.com/ejgallego/coq-lsp/pull/696
- https://github.com/coq-community/coq-dpdgraph/pull/132
- https://github.com/coq-tactician/coq-tactician/pull/82
- https://github.com/HoTT/Coq-HoTT/pull/1951

**Review by ppedrot — APPROVED** [review]

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 8 above)
Comment (ejgallego) [reply doc/sphinx/proof-engine/vernacular-commands.rst:665]: Note that this is a problem for some users, as there are findlib libraries that don't contain a `.`.

So IMHO, this is a RC bug.

File: doc/changelog/08-vernac-commands-and-options/18385-remove_legacy_build_mode.rst [hunk 10]
```diff
@@ -0,0 +1,4 @@
+- **Removed:**
+  "legacy" (non-findlib) loading mode for plugins in :cmd:`Declare ML Module`
+  (`#18385 <https://github.com/coq/coq/pull/18385>`_,
+  by Gaëtan Gilbert).
```
Comment (ejgallego) [inline doc/changelog/08-vernac-commands-and-options/18385-remove_legacy_build_mode.rst:4]: @SkySkimmer shouldn't this include me too?
